services/wallet/bridge/hop.go

```diff
@@ -0,0 +1,378 @@
+		txOpts,
+		big.NewInt(int64(sendArgs.ChainID)),
+		sendArgs.Recipient,
+		sendArgs.Amount.ToInt(),
+		big.NewInt(0),
+		deadline,
+		common.HexToAddress("0x0"),
+		big.NewInt(0),
+	)
+
+	if err != nil {
+		return hash, err
+	}
+	return types.Hash(tx.Hash()), nil
+}
+
+func (s *HopBridge) swapAndSend(chainID uint64, sendArgs *HopTxArgs, verifiedAccount *account.SelectedExtKey) (hash types.Hash, err error) {
+	ammWrapper, err := s.contractMaker.NewHopL2AmmWrapper(chainID, sendArgs.Symbol)
+	if err != nil {
+		return hash, err
+	}
+
+	txOpts := sendArgs.ToTransactOpts(getSigner(chainID, sendArgs.From, verifiedAccount))
+	txOpts.Value = (*big.Int)(sendArgs.Amount)
+	now := time.Now()
+	deadline := big.NewInt(now.Unix() + 604800)
+	tx, err := ammWrapper.SwapAndSend(
+		txOpts,
+		big.NewInt(int64(sendArgs.ChainID)),
+		sendArgs.Recipient,
+		sendArgs.Amount.ToInt(),
+		sendArgs.BonderFee.ToInt(),
+		big.NewInt(0),
+		deadline,
+		big.NewInt(0),
+		deadline,
+	)
+
+	if err != nil {
+		return hash, err
+	}
+
+	return types.Hash(tx.Hash()), nil
+}
+
+func (h *HopBridge) CalculateBonderFees(from, to *params.Network, token *token.Token, amountIn *big.Int, nativeTokenPrice, tokenPrice float64, gasPrice *big.Float) (*big.Int, error) {
+	amount := new(big.Float).SetInt(amountIn)
+	totalFee := big.NewFloat(0)
+	destinationTxFee, err := h.getDestinationTxFee(from, to, nativeTokenPrice, tokenPrice, gasPrice)
+	if err != nil {
+		return nil, err
+	}
+
+	bonderFeeRelative, err := h.getBonderFeeRelative(from, to, amount, token)
+	if err != nil {
+		return nil, err
+	}
+	if from.Layer != 1 {
+		adjustedBonderFee, err := h.calcFromHTokenAmount(to, bonderFeeRelative, token.Symbol)
+		if err != nil {
+			return nil, err
+		}
+		adjustedDestinationTxFee, err := h.calcToHTokenAmount(to, destinationTxFee, token.Symbol)
+		if err != nil {
+			return nil, err
+		}
+
+		bonderFeeAbsolute := h.getBonderFeeAbsolute(tokenPrice)
+		if adjustedBonderFee.Cmp(bonderFeeAbsolute) == -1 {
+			adjustedBonderFee = bonderFeeAbsolute
+		}
+
+		totalFee.Add(adjustedBonderFee, adjustedDestinationTxFee)
+	}
+	res, _ := new(big.Float).Mul(totalFee, big.NewFloat(math.Pow(10, float64(token.Decimals)))).Int(nil)
+	return res, nil
+}
+
+func (h *HopBridge) CalculateFees(from, to *params.Network, token *token.Token, amountIn *big.Int, nativeTokenPrice, tokenPrice float64, gasPrice *big.Float) (*big.Int, *big.Int, error) {
+	bonderFees, err := h.CalculateBonderFees(from, to, token, amountIn, nativeTokenPrice, tokenPrice, gasPrice)
+	if err != nil {
+		return nil, nil, err
+	}
+	amountOut, err := h.amountOut(from, to, new(big.Float).SetInt(amountIn), token.Symbol)
+	if err != nil {
+		return nil, nil, err
+	}
+	amountOutInt, _ := amountOut.Int(nil)
+
+	return bonderFees, new(big.Int).Add(
+		bonderFees,
+		new(big.Int).Sub(amountIn, amountOutInt),
+	), nil
+}
+
+func (h *HopBridge) calcToHTokenAmount(network *params.Network, amount *big.Float, symbol string) (*big.Float, error) {
```

Comment:
this name of function is not understandable to me

Reply:
This is the algorithm provided by hop, I will provide a link to it

Reply:
hey no I meant the name calcToHTokenAmount? do we need to keep it this way?